once/__init__.py

```diff
@@ -46,78 +51,151 @@ def _wrapped_function_type(func: collections.abc.Callable) -> _WrappedFunctionTy
+            once_base: _OnceBase = once_factory()
+            async with once_base.async_lock:
+                if not once_base.called:
+                    once_base.return_value = _iterator_wrappers.AsyncGeneratorWrapper(
+                        func, *args, **kwargs
+                    )
+                    once_base.called = True
+                return_value = once_base.return_value
+            next_value = None
+            iterator = return_value.yield_results()
+            while True:
+                try:
+                    next_value = yield await iterator.asend(next_value)
+                except StopAsyncIteration:
+                    return
+
+    elif fn_type == _WrappedFunctionType.ASYNC_FUNCTION:
+
+        async def wrapped(*args, **kwargs) -> typing.Any:
+            once_base: _OnceBase = once_factory()
+            async with once_base.async_lock:
+                if not once_base.called:
+                    once_base.return_value = await func(*args, **kwargs)
+                    once_base.called = True
+                return once_base.return_value
+
+    elif fn_type == _WrappedFunctionType.SYNC_FUNCTION:
+
+        def wrapped(*args, **kwargs) -> typing.Any:
+            once_base: _OnceBase = once_factory()
+            with once_base.lock:
+                if not once_base.called:
+                    once_base.return_value = func(*args, **kwargs)
+                    once_base.called = True
+                return once_base.return_value
+
+    elif fn_type == _WrappedFunctionType.SYNC_GENERATOR:
+
+        def wrapped(*args, **kwargs) -> typing.Any:
+            once_base: _OnceBase = once_factory()
+            with once_base.lock:
+                if not once_base.called:
+                    once_base.return_value = _iterator_wrappers.GeneratorWrapper(
+                        func, *args, **kwargs
+                    )
+                    once_base.called = True
+                iterator = once_base.return_value
+            yield from iterator.yield_results()
+
+    else:
+        raise NotImplementedError()
+
+    functools.update_wrapper(wrapped, func)
+    return wrapped
+
+
+class _PerItemOnceMap:
+    def __init__(self, is_async: bool) -> None:
+        self.is_async = is_async
+        self.per_item_onces: weakref.WeakKeyDictionary[
+            typing.Any, _OnceBase
+        ] = weakref.WeakKeyDictionary()
+        self.lock = threading.Lock()
+
+    def __getitem__(self, obj: typing.Any) -> _OnceBase:
+        with self.lock:
+            if (obj_once := self.per_item_onces.get(obj)) is None:
+                obj_once = _OnceBase(self.is_async)
+                self.per_item_onces[obj] = obj_once
+        return obj_once
 
-        functools.update_wrapper(wrapped, func)
-        return wrapped
 
+def _once_factory(is_async: bool, per_thread: bool) -> _ONCE_FACTORY_TYPE:
+    if per_thread:
+        per_thread_onces = _PerItemOnceMap(is_async)
```

Comment:
Nit: _PerItemOnceMap is only used here and is basically reimplementing threading.local()

I think this could just be:
```python
if per_thread:
    per_thread_onces = threading.local()
    # Part of me realllyyyyy wonders if we could do
    # per_thread_onces.__dict__ = defaultdict(lambda: _OnceBase(is_async)) | per_thread_onces.__dict__
    per_thread_onces.__dict__.set_default("once", None)
    def _get_once_per_thread():
        if per_thread_onces.once:
              return per_thread_onces.once
        per_thread_onces.once = _OnceBase(is_async)
        return per_thread_onces.once
    return _get_once_per_thread
```

Reply:
done!